src/devices/machine/qvt102.cpp

```diff
@@ -0,0 +1,692 @@
+	PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_TAB)        PORT_CHAR(9)
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("P2_7")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_DEL)        PORT_CHAR(UCHAR_MAMEKEY(DEL))
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_HOME)       PORT_CHAR(UCHAR_MAMEKEY(HOME))      PORT_NAME("Clear / Home")
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_RIGHT)      PORT_CHAR(UCHAR_MAMEKEY(RIGHT))     PORT_NAME("\xe2\x86\x92") // →
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_CLOSEBRACE) PORT_CHAR(']') PORT_CHAR('}')
+	PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_1)          PORT_CHAR('1') PORT_CHAR('!')
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("SPECIAL")
+	PORT_BIT(0x01, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_LSHIFT)   PORT_CODE(KEYCODE_RSHIFT)   PORT_CHAR(UCHAR_SHIFT_1) // ⇧
+	PORT_BIT(0x02, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_LCONTROL) PORT_CODE(KEYCODE_RCONTROL) PORT_CHAR(UCHAR_MAMEKEY(LCONTROL)) PORT_NAME("Ctrl")
+	PORT_BIT(0x04, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CODE(KEYCODE_CAPSLOCK)                             PORT_CHAR(UCHAR_MAMEKEY(CAPSLOCK))
+
+	PORT_START("JUMPER")
+	PORT_DIPNAME(0x01, 0x00, "EIA DSR (W1)")
+	PORT_DIPSETTING(   0x00, DEF_STR( Off ))
+	PORT_DIPSETTING(   0x01, DEF_STR( On ))
+	PORT_DIPNAME(0x02, 0x02, "EIA DCD (W2)")
+	PORT_DIPSETTING(   0x00, DEF_STR( Off ))
+	PORT_DIPSETTING(   0x02, DEF_STR( On ))
+	PORT_DIPNAME(0x04, 0x04, "EIA DTR (W3)")
+	PORT_DIPSETTING(   0x00, DEF_STR( Off ))
+	PORT_DIPSETTING(   0x04, DEF_STR( On ))
+	PORT_DIPNAME(0x08, 0x00, "EIA RTS (W4)")
+	PORT_DIPSETTING(   0x00, DEF_STR( Off ))
+	PORT_DIPSETTING(   0x08, DEF_STR( On ))
+	PORT_DIPNAME(0x30, 0x00, "Character Set")
+	PORT_DIPSETTING(   0x00, "US (W17)")
+	PORT_DIPSETTING(   0x10, "GM (W18)")
+	PORT_DIPSETTING(   0x20, "UK (W19)")
+	PORT_DIPNAME(0x40, 0x00, "Attribute Code Intensity")
+	PORT_DIPSETTING(   0x00, "Half (W21)")
+	PORT_DIPSETTING(   0x40, "Full (W22)")
+INPUT_PORTS_END
+
+
+//**************************************************************************
+//  KEYBOARD
+//**************************************************************************
+
+// 7------- unused
+// -654---- row select/special keys
+// ----3--- unknown
+// -----21- unused
+// -------0 speaker
+
+READ8_MEMBER(qvt102_device::mcu_bus_r)
+{
+	return m_kbd_bus;
+}
+
+WRITE8_MEMBER(qvt102_device::mcu_bus_w)
+{
+	m_speaker->level_w(BIT(data, 0));
+	m_kbd_bus = data;
+}
+
+READ_LINE_MEMBER(qvt102_device::mcu_t0_r)
+{
+	int state = 1;
+
+	if (BIT(m_kbd_bus, 4) == 0) state = BIT(m_keys_special->read(), 0);
+	if (BIT(m_kbd_bus, 5) == 0) state = BIT(m_keys_special->read(), 1);
+	if (BIT(m_kbd_bus, 6) == 0) state = BIT(m_keys_special->read(), 2);
```

Comment:
Does this really have a priority mechanism?  If it doesn't, it should be `state &=` on each line; if it really does have a priority mechanism it should be `else if` to make it clearer.

Reply:
This and other some other queries are for the old code, but let me take a look.